experimental/sql/dialects/rel_ssa.py

```diff
@@ -0,0 +1,320 @@
+# Licensed under the Apache License v2.0 with LLVM Exceptions.
+# See https://llvm.org/LICENSE.txt for license information.
+# SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+from dataclasses import dataclass
+from typing import Any, Union, List
+from xdsl.ir import Block, Region, Operation, SSAValue, ParametrizedAttribute, Data, MLContext, Attribute
+from xdsl.dialects.builtin import StringAttr, ArrayAttr, ArrayOfConstraint, IntegerAttr
+from xdsl.irdl import AttributeDef, OperandDef, ResultDef, RegionDef, SingleBlockRegionDef, irdl_attr_definition, irdl_op_definition, ParameterDef, AnyAttr, VarOperandDef, builder
+
+# This file contains the relational SSA dialect, a dialect that expresses
+# relational queries using SSA. The dialect consists of two types of operations:
+# Expressions and Operators. Operators work on Bags whereas expressions work on
+# single values and are used to specify the exact behavior of operators. The
+# Column and Yield Expressions bridge the gap to convert from one to the other.
+
+#===------------------------------------------------------------------------===#
+# Data types
+#===------------------------------------------------------------------------===#
+
+
+@irdl_attr_definition
+class DataType(ParametrizedAttribute):
+  """
+  Models a datatype in a relational SSA query.
+  """
+  name = "rel_ssa.datatype"
+
+
+@irdl_attr_definition
+class Int32(DataType):
+  """
+  Models a int32 type in a relational SSA query.
+
+  Example:
+
+  ```
+  !rel_ssa.int32
+  ```
+  """
+  name = "rel_ssa.int32"
+
+
+@irdl_attr_definition
+class String(DataType):
+  """
+  Models a string type in a relational SSA query, that can be either nullable or
+  not.
+
+  Example:
+
+  ```
+  !rel_ssa.string<0: !i1>
+  ```
+  """
+  name = "rel_ssa.string"
+
+  # TODO: redefine nullable as a property of all fields
+  nullable = ParameterDef(IntegerAttr)
+
+  @staticmethod
+  @builder
+  def get(val: Union[int, IntegerAttr]) -> 'String':
+    if isinstance(val, IntegerAttr):
+      return String([val])
+    return String([IntegerAttr.from_int_and_width(val, 1)])
```

Comment:
This constructs a String from an IntegerAttr or from an int. Whether or not the string is nullable is decided by the int in one case or by the IntegerAttr (that internally has a value field).

In essence, the union type and the `isinstance` were a try to have an API that looks simple (always use `get(...)` to build something), while still being expressive (can use int and IntegerAttr to build). In the end, I feel that this kind of code should be removed, since it is quite ugly. The approach of having two functions `from_int` and `from_integerattr` would probably work better or using the mutipledispach module to dispatch to the right thing according to the type.

And again, I actually want a `Bool` but that is not supported yet, so i1 is used in this prototype.

Reply:
Sorry for chiming in late but it might also help to rename `val`, for example, to `is_nullable`.